docs/project_proposal.md

```diff
@@ -0,0 +1,116 @@
+# Starlark Language Server Proposal
+
+Do you wish that your editor/IDE provided fancy capabilities for [Starlark] (e.g., auto complete, go
+to definition, hover information)? If so, read on and help us get to our funding goal to make this
+project a reality.
+
+| Quick Facts  |                                                                                                                                                          |
+| ------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| What         | Implement a language server for [Starlark].                                                                                                              |
+| Why          | Fancy features like go to definition, hover information and auto complete in your favorite editor/IDE. (see [Language Server Protocol] for more details) |
+| Who          | [Chuck Grindel]                                                                                                                                          |
+| When         | As soon as we meet our funding goals.                                                                                                                    |
+| Language     | Implement the server using [Go].                                                                                                                         |
+| License      | [Apache 2.0]                                                                                                                                             |
+| GitHub Owner | <https://github.com/bazel-contrib>                                                                                                                       |
+| Contribute   | Help us reach our funding goals by [contributing].                                                                                                       |
+
+## 💻 What is a language server?
+
+A language server is a command-line executable that implements the [Language Server Protocol] (LSP).
+It interacts with a client in your editor/IDE to provide the smarts that every developer loves and
+wishes that they had for [Starlark].
+
+## 📝 The Plan
+
+The plan consists of four phases:
+
+1. Initial server implementation plus support for go to definition and find references.
+1. Support hover information.
+1. Support auto completion.
+1. Support signature help.
+
+Once we meet our initial funding goals, other features that are on the roadmap are support
+for
+
+- linting
+- formatting
+- code lens (provide contextual information and actions)
```

Comment:
When I tried to do this, one of the main roadblocks was resolving third party rulesets to the correct versions in a non-blocking way.
As far as I understand it, one would need to either:

- Have their own execution engine for repository rules, or
- Rely on `bazel sync` calls to resolve workspace dependencies.

The second approach is the obvious one, but it has the drawback of blocking bazel for potentially long stretches. So, say that one:

- opens a bazel workspace in VScode -> initial bazel sync starts.
- Makes a small edit to one file
- Wants to run `bazel test //...` to test everything because they have remote execution and they expect only one test to be run.

In that case, would we block until the bazel sync is complete? Preempt the sync command? Something else?

Reply:
Just like intellij needs to index a Maven project on startup(including download the maven dependencies), we gonna have to `index` the Bazel project with the LSP to provide a responsive UX (read "low latency").

This index could include:
- Prefetching external deps onto disk (using `bazel query //<target>`)
- Warm up Bazel JVM daemon
- Build up reference map in-memory

with future configurations that (1) allow doing those lazily or in a background async process and (2) do things in smaller, fine-grained chunks.

But I think it's reasonable for an early implementation to assume a warm Bazel JVM with all external deps already available on disk. We could iterate over time based on the telemetry data I proposed above.

Reply:
I planned to accept the cost of indexing the world for the first pass of the language server. 

🤔 Thinking out loud, can we leverage the Bazel cache to cache the indexing results? If we wrote an aspect that defines actions to do the indexing, the results could be cached. The downside is that the indexing actions would be blocked or block the user's activity. I know that `rules_xcodeproj` handles this by creating separate output paths for indexing and more mainline activity.